lib/mpl_toolkits/tests/test_axes_grid.py

```diff
@@ -39,12 +40,13 @@ def test_imagegrid_cbar_mode_edge():
         ax3.imshow(np.abs(arr), cmap='jet')
         ax4.imshow(np.arctan2(arr.imag, arr.real), cmap='hsv')
 
-        # In each row/column, the "first" colorbars must be overwritten by the
-        # "second" ones.  To achieve this, clear out the axes first.
         for ax in grid:
-            ax.cax.cla()
+            # We don't want to share the norms between the colorbar axes
+            # so create a deepcopy for each image
+            sm = mpl.cm.ScalarMappable(norm=copy.deepcopy(ax.images[0].norm),
+                                       cmap=ax.images[0].get_cmap())
             cb = ax.cax.colorbar(
-                ax.images[0],
+                sm,
```

Comment:
I think @anntzer worked on this last, trying to make these colorbar-like objects the same as main-library colorbars.

Reply:
This still feels like there is something different about `axes_grid` that I am missing and we shouldn't need to deepcopy the norm on the mappable... My guess is there is possibly a locator or something similar attached to `ax.cax` that needs to be cleared/reset on the `ax.cax.cla()` call above, but I haven't been able to find it while debugging and have spent quite a lot of time on this.

Is there a way to step through a debugger session while keeping the plot window open to see how things move around while stepping into functions?

Reply:
Not that I know of, but I never use a debugger (probably a bad idea).

Reply:
I'm still deeply confused by this test.  It seems that it intentionally over-rides the axes so I guess the point is to show that 'cax' for both of the axes in a given row / column?


![imagegrid_cbar_mode](https://user-images.githubusercontent.com/199813/119433912-26d5a380-bce5-11eb-9f21-62c25dbdf599.png)

I am not sure what the expected behavior here actual is to sort out how big of a change this is...

Reply:
I am very confused as well. I _think_ what is happening is that on the first pass of colorbar addition to the cax, some locators/formatters get updated and set due to the new callback. Then on the second addition of the colorbar, that state is still hanging around. Rather than deepcopying, I switched the test to remove the callback from the colorbar.

The difference with/without the final commit is that the second colorbar axis has ticklabels showing up. I'm not exactly sure where this state creeps in, so this is a hard one to trace down.
![imagegrid_cbar_mode-failed-diff](https://user-images.githubusercontent.com/12417828/119525032-662edf00-bd3b-11eb-83b2-de52538f7cc4.png)

Reply:
@anntzer worked on this last in imageGrid, so it would be very helpful if he could describe what is going on here.

Reply:
From a quick look, this seems to arise (at least to some degree) because remove_text doesn't know how to hand axisartist-axes, which require a different API for removing ticklabels (because because).
(It's not clear to me why it worked before, though.)

Reply:
Thanks for looking into this @anntzer! I'll try and dig a little more into the axisartist/remove_text and see if there is a quick fix in there. Otherwise, I think a better workaround would likely be to move my fix in the test internal to the axes_grid colorbar routine and put a TODO in there for adding that capability back at a later point.

Reply:
OK, I figured out this problem.  This is basically due to the interaction between two points.

First, as discussed during the call, cla() doesn't call Colorbar.remove(), which means that the callback of the first image doesn't get deleted, which it should.  Making cla() actually go through each and every artist to call cla() on it may be too expensive, but at least perhaps we can check if there's any colorbar on the axes and properly remove() them.

Still, even knowing that, it was unclear why the first image would overwrite the second colorbar.  The reason is because the sequence of events was the following:
- instantiate first image, create associated colorbar.
- instantiate second image, create associated colorbar.
- actually walk the draw tree and draw the images... which has the effect of calling _ImageBase._make_image, which contains the following snippet:
```python
                # we have re-set the vmin/vmax to account for small errors
                # that may have moved input values in/out of range
                s_vmin, s_vmax = vrange
                if isinstance(self.norm, mcolors.LogNorm) and s_vmin <= 0:
                    # Don't give 0 or negative values to LogNorm
                    s_vmin = np.finfo(scaled_dtype).eps
                with cbook._setattr_cm(self.norm,
                                       vmin=s_vmin,
                                       vmax=s_vmax,
                                       ):
                    output = self.norm(resampled_masked)
```
so here s_vmin and s_vmax were recomputed to take interpolation into account.  Unfortunately, this means they may be slightly different due to the rescaling and unscaling, which can be checked by printing `s_vmin - A.min(), s_vmax - A.max()`.  One then sees that there's indeed a tiny (1e-16) difference for the first image's norm, which means it'll say, oh, I need to redraw the colorbar; but there's no difference(!) for the second image's norm, so it doesn't redraw itself.  A quickfix to confirm that this works is that forcing `s_vmin = A.min(); s_vmax = A.max()` here fixes the issue (but would cause other problems with over/under interaction with resampling; I suspect that the "correct" fix is to temporarily disable emission of callbacks by `self.norm` just for the call to `_setattr_cm`.

That was nasty :p

Reply:
That looks like it did the trick. I disconnected from the callback before this and reconnected after it. I wonder if more fundamentally there should be something within the context manager to remove the callbacks on enter/exit...

Reply:
I guess we could add a `disabling_callbacks()` context manager on CallbackRegistry?
```python
with self.norm.callbacks.disabling_callbacks(), cbook._setattr_cm(self.norm, ...): ...
```

Reply:
Yeah, that looks pretty clean to me. I'm not sure how much more use it would have in more than this one situation though, so is it worth adding or not...

Lets keep it as a separate issue and implement it later.

Reply:
Rebased and updated this with the blocked callback feature now.